socs/agents/acu/avoidance.py

```diff
@@ -0,0 +1,685 @@
+            # Add the selected trajectory in green.
+            selected = self.select_move(all_moves)[0]
+            if selected is not None:
+                traj = selected['moves'].get_traj()
+                segments = self._azel_pix(*traj, round=True, segments=True)
+                for ax in axes:
+                    for j, i in segments:
+                        ax.plot(i, j, color='green')
+
+            pl.savefig(plot_file)
+        return all_moves
+
+    def find_escape_paths(self, az0, el0, t=None,
+                          debug=False):
+        """Design and analyze a number of different paths that move from (az0,
+        el0) to a sun safe position.  Return the list, for further
+        processing and choice.
+
+        """
+        if t is None:
+            t = self._now()
+
+        az_cands = []
+        _az = math.ceil(self.policy['min_az'] / 180) * 180
+        while _az <= self.policy['max_az']:
+            az_cands.append(_az)
+            _az += 180.
+
+        # Clip el0 into the allowed range.
+        el0 = np.clip(el0, self.policy['min_el'], self.policy['max_el'])
+
+        # Preference is to not change altitude; but allow for lowering.
+        n_els = math.ceil(el0 - self.policy['min_el']) + 1
+        els = np.linspace(el0, self.policy['min_el'], n_els)
+
+        path = None
+        for el1 in els:
+            paths = [self.analyze_paths(az0, el0, _az, el1, t=t, dodging=False)
+                     for _az in az_cands]
+            best_paths = [self.select_move(p)[0] for p in paths]
+            best_paths = [p for p in best_paths if p is not None]
+            if len(best_paths):
+                path = self.select_move(best_paths)[0]
+                if debug:
+                    cands, _ = self.select_move(best_paths, raw=True)
+                    return cands
+            if path is not None:
+                return path
+
+        return None
+
+    def select_move(self, moves, raw=False):
+        """Given a list of possible "moves", select the best one.
+        The "moves" should be like the ones returned by
+        ``analyze_paths``.
+
+        The best move is determined by first screening out dangerous
+        paths (ones that pass close to Sun, move closer to Sun
+        unnecessarily, violate axis limits, etc.) and then identifying
+        paths that minimize danger (distance to Sun; Sun time) and
+        path length.
+
+        If raw=True, a debugging output is returned; see code.
+
+        Returns:
+          best_move (dict): The element of moves that is safest.  If
+            no safe move was found, None is returned.
+          decisions (list): The items in this list are dicts that
+            correspond one-to-one with the entries in moves.  Each
+            decision dict has entries 'rejected' (True or False) and
+            'reason' (string description of why the move was rejected
+            outright).
```

Comment:
The autodoc formatting for this is not quite right. It's rendering like:
![image](https://github.com/simonsobs/socs/assets/7691438/6f022048-da72-4b6a-8f1d-19ac59734412)

The syntax is a bit different than for arguments, it's:
```
Returns:
    type: Description text.
    
    Multiple lines can be used indented to the same level as the first line.
```

So something like this would detect the type properly:
```
        Returns:
          (dict, list): (best_move, decisions)

          ``best_move`` -- the element of moves that is safest.  If no safe
          move was found, None is returned.

          ``decisions`` -- The items in this list are dicts that correspond
          one-to-one with the entries in moves.  Each decision dict has entries
          'rejected' (True or False) and 'reason' (string description of why
          the move was rejected outright).
```

This renders like:
![image](https://github.com/simonsobs/socs/assets/7691438/4cc00247-1b12-4493-bb4b-e5435ce401e1)

Reply:
Ah, great ... will do that.